wsmaster/che-core-api-installer/src/main/java/org/eclipse/che/api/installer/server/jpa/JpaInstallerDao.java

```diff
@@ -0,0 +1,176 @@
+import org.eclipse.che.api.installer.server.impl.InstallerFqn;
+import org.eclipse.che.api.installer.server.model.impl.InstallerImpl;
+import org.eclipse.che.api.installer.server.spi.InstallerDao;
+import org.eclipse.che.api.installer.shared.model.Installer;
+import org.eclipse.che.core.db.DBInitializer;
+import org.eclipse.che.core.db.jpa.DuplicateKeyException;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.persistence.EntityManager;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * @author Anatolii Bazko
+ */
+@Singleton
+public class JpaInstallerDao implements InstallerDao {
+
+    private final Provider<EntityManager> managerProvider;
+
+    @Inject
+    public JpaInstallerDao(@SuppressWarnings("unused") DBInitializer dbInitializer,
+                           Provider<EntityManager> managerProvider) {
+        this.managerProvider = managerProvider;
+    }
+
+    @Override
+    public void create(InstallerImpl installer) throws InstallerException {
+        requireNonNull(installer, "Required non-null installer");
+        try {
+            doCreate(installer);
+        } catch (DuplicateKeyException x) {
+            throw new InstallerConflictException(
+                    format("Installer with such fqn '%s:%s' already exists", installer.getId(), installer.getVersion()));
+        } catch (RuntimeException x) {
+            throw new InstallerException(x.getMessage(), x);
+        }
+    }
+
+    @Override
+    public void update(InstallerImpl installer) throws InstallerException {
+        requireNonNull(installer, "Required non-null update");
+        try {
+            doUpdate(installer);
+        } catch (RuntimeException x) {
+            throw new InstallerException(x.getMessage(), x);
+        }
+    }
+
+    @Override
+    public void remove(InstallerFqn fqn) throws InstallerException {
+        requireNonNull(fqn, "Required non-null fqn");
+        try {
+            doRemove(fqn);
+        } catch (RuntimeException x) {
+            throw new InstallerException(x.getMessage(), x);
+        }
+    }
+
+    @Override
+    @Transactional
+    public InstallerImpl getByFqn(InstallerFqn fqn) throws InstallerException {
+        requireNonNull(fqn, "Required non-null fqn");
+
+        try {
+            InstallerImpl installer = managerProvider.get().find(InstallerImpl.class, fqn);
+            if (installer == null) {
+                throw new InstallerNotFoundException(format("Installer with fqn '%s' doesn't exist", fqn));
+            }
+            return new InstallerImpl(installer);
+        } catch (RuntimeException e) {
+            throw new InstallerException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    @Transactional
+    public Page<InstallerImpl> getVersions(String id, int maxItems, long skipCount) throws InstallerException {
```

Comment:
Manager and Service have the method with the same but do something different. Manager and Service return all versions of installer with id, while dao return installer with all versions with id. It can be confused.
Should we really implement it in such way or we can implement it in the same way?